File: 003 Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
    	int len = s.length();
    	if (len == 0) return 0;
    	int start = 0, end = 0, max = 1, sublen = 0;
    	Map<Character, Integer> map = new HashMap<Character, Integer>();
    	while (end < len) {
    		if (!map.containsKey(s.charAt(end))){
    			map.put(s.charAt(end++), 1);
    			sublen++;
    			max = Math.max(sublen, max);
    		} else { 
    			map.remove(s.charAt(start++));
    			sublen--;
    		}
    	}
    	return max;
    }
}

File: 003 Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i = res = 0
        dic = {}
        
        for j in range(len(s)):
            if s[j] in dic and i <= dic[s[j]]:
                i = dic[s[j]] + 1
            else:
                res = max(res, j - i + 1)
            dic[s[j]] = j
            
        return res